DataFormats/GEMRecHit/src/GEMCSCSegment.cc

```diff
@@ -0,0 +1,124 @@
+GEMCSCSegment::GEMCSCSegment(const CSCSegment* csc_segment, const std::vector<const GEMRecHit*> gem_rhs, LocalPoint origin, LocalVector direction, AlgebraicSymMatrix errors, double chi2) : 
+
+  RecSegment(buildDetId(csc_segment->cscDetId())),
+  theOrigin(origin), 
+  theLocalDirection(direction), theCovMatrix(errors), theChi2(chi2) {
+
+  for(unsigned int i=0; i<gem_rhs.size(); ++i) {
+    theGEMRecHits.push_back((*gem_rhs[i]));
+  }
+  theCSCSegment = *csc_segment;
+  // LogDebug
+  edm::LogVerbatim("GEMCSCSegment")<< "[GEMCSCSegment :: ctor] CSCDetId: " << csc_segment->cscDetId() << " CSC RecHits: " <<csc_segment->specificRecHits().size()
+				   << " GEM RecHits: " << gem_rhs.size()<<"\n" //  << " Fit chi2: "<<chi2<<" Position: "<<origin<<" Direction: "<<direction
+				   << "    CSC Segment Details: \n"<<*csc_segment<<"\n"
+				   << " GEMCSC Segment Details: \n"<<*this<<"\n"
+				   << "[GEMCSCSegment :: ctor] ------------------------------------------------------------";
+}
+
+
+GEMCSCSegment::~GEMCSCSegment() {}
+
+
+std::vector<const TrackingRecHit*> GEMCSCSegment::recHits() const{
+
+  std::vector<const TrackingRecHit*> pointersOfRecHits;
+  for (std::vector<GEMRecHit>::const_iterator irh = theGEMRecHits.begin(); irh!=theGEMRecHits.end(); ++irh) {
+    pointersOfRecHits.push_back(&(*irh));
+  }
+  for (std::vector<CSCRecHit2D>::const_iterator irh = theCSCSegment.specificRecHits().begin(); irh!=theCSCSegment.specificRecHits().end(); ++irh) {
+    pointersOfRecHits.push_back(&(*irh));
+  }
+  return pointersOfRecHits;
+}
+
+std::vector<TrackingRecHit*> GEMCSCSegment::recHits() {
+
+  std::vector<TrackingRecHit*> pointersOfRecHits;
+  for (std::vector<GEMRecHit>::iterator irh = theGEMRecHits.begin(); irh!=theGEMRecHits.end(); ++irh) {
+    pointersOfRecHits.push_back(&(*irh));
+  }
+  /*
+  for (std::vector<CSCRecHit2D>::const_iterator irh = theCSCSegment.specificRecHits().begin(); irh!=theCSCSegment.specificRecHits().end(); ++irh) {
+    pointersOfRecHits.push_back(&(*irh));
+    }*/
+  return pointersOfRecHits;
+}
+
+
+LocalError GEMCSCSegment::localPositionError() const {
+  return LocalError(theCovMatrix[2][2], theCovMatrix[2][3], theCovMatrix[3][3]);
+}
+
+
+LocalError GEMCSCSegment::localDirectionError() const {
+  return LocalError(theCovMatrix[0][0], theCovMatrix[0][1], theCovMatrix[1][1]); 
+}
+
+
+AlgebraicVector GEMCSCSegment::parameters() const {
+  // For consistency with DT, CSC and what we require for the TrackingRecHit interface,
+  // the order of the parameters in the returned vector should be (dx/dz, dy/dz, x, z)
+  
+  AlgebraicVector result(4);
+
+  result[0] = theLocalDirection.x()/theLocalDirection.z();
+  result[1] = theLocalDirection.y()/theLocalDirection.z();    
+  result[2] = theOrigin.x();
+  result[3] = theOrigin.y();
+
+  return result;
+}
+
+
+AlgebraicMatrix GEMCSCSegment::projectionMatrix() const {
```

Comment:
Hi Carl

Sorry, I had been taken by some other work the past 10 days ... Now I am
back on the job and I would like some clarification on your
statement/position here...

1) you are asking this for the projectionMatrix() and not for the
parameters() method? To be consistent, it should be done for both I would
guess (but i am not an expert here)
2) applying it for the projectionMatrix() it complains about it:

/afs/
cern.ch/cms/sw/ReleaseCandidates/vol0/slc6_amd64_gcc491/cms/cmssw/CMSSW_7_5_X_2015-05-26-2300/src/DataFormats/TrackingRecHit/interface/TrackingRecHit.h:91:27:
error:   overriding 'virtual AlgebraicMatrix
TrackingRecHit::projectionMatrix() const'
   virtual AlgebraicMatrix projectionMatrix() const = 0;

seems that the method overrides the virtual method from the TrackingRecHit
from which it inherits... Does that mean that actually there we should also
change it to return a pointer to a AlgebraicMatrix?

Just for completeness ... it is good to change this kind of things, but
this means that if we change it for TrackingRecHit, then also CSC and DT
segment dataformat needs to be adapted, because they also return just
AlgebraicMatrices and not a const refernce to it.

Can you clarify a bit what I should do, or did I misunderstand the code (it
is a likely option, but I would be happy if you could say a word about it)
Thanks & Greets
Piet


On Mon, May 18, 2015 at 1:55 AM, Carl Vuosalo <notifications@github.com>
wrote:

> In DataFormats/GEMRecHit/src/GEMCSCSegment.cc
> <https://github.com/cms-sw/cmssw/pull/9083#discussion_r30476449>:
>
> > +AlgebraicVector GEMCSCSegment::parameters() const {
> > +  // For consistency with DT, CSC and what we require for the TrackingRecHit interface,
> > +  // the order of the parameters in the returned vector should be (dx/dz, dy/dz, x, z)
> > +
> > +  AlgebraicVector result(4);
> > +
> > +  result[0] = theLocalDirection.x()/theLocalDirection.z();
> > +  result[1] = theLocalDirection.y()/theLocalDirection.z();
> > +  result[2] = theOrigin.x();
> > +  result[3] = theOrigin.y();
> > +
> > +  return result;
> > +}
> > +
> > +
> > +AlgebraicMatrix GEMCSCSegment::projectionMatrix() const {
>
> Return a const reference to the matrix to avoid unnecessary copies.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9083/files#r30476449>.
>



-- 
Piet Verwilligen

INFN -- Sezione di Bari
Via E. Orabona 4
I-70125 Bari, Italy
Phone: +39 345 74 70 642